controllers/users.js

```diff
@@ -0,0 +1,78 @@
+const jwt = require('jsonwebtoken');
+const bcrypt = require('bcryptjs');
+const User = require('../models/user');
+const NotFoundError = require('../errors/notFoundError');
+const { JWT_SECRET_KEY } = require('../config');
+
+
+module.exports.createUser = (req, res) => {
+  const {
+    name, about, avatar, email, password,
+  } = req.body;
+
+  bcrypt.hash(password, 10)
+    .then((hash) => User.create({
+      name, about, avatar, email, password: hash,
+    }))
+    .then((user) => user.omitPrivate())
+    .then((user) => res.send({ data: user }))
+    .catch((err) => {
+      if (err.code === 11000) {
+        res.status(409).send({ message: 'Пользователь с таким email уже существует' });
+      } else {
+        res.status(500).send({ message: 'Произошла ошибка' });
+      }
+    });
+};
+
+module.exports.getUsers = (req, res) => {
+  User.find({})
+    .then((user) => res.send({ data: user }))
+    .catch(() => res.status(500).send({ message: 'Произошла ошибка' }));
+};
+
+
+module.exports.getUser = (req, res) => {
+  const { userId } = req.params;
+
+  User.findById(userId)
+    .orFail(() => new NotFoundError('Пользователь не найден'))
+    .then((user) => res.send({ data: user }))
+    .catch((err) => res.status(err.statusCode || 500).send({ message: 'Произошла ошибка', err: err.message }));
+};
+
+
+module.exports.updateProfile = (req, res) => {
+  const userId = req.user._id;
+  const { name, about } = req.body;
+
+  User.findByIdAndUpdate(userId, { name, about }, { new: true, runValidators: true })
+    .then((user) => res.send({ newData: user }))
+    .catch(() => res.status(500).send({ message: 'Произошла ошибка' }));
+};
+
+module.exports.updateProfileAvatar = (req, res) => {
+  const userId = req.user._id;
+  const { avatar } = req.body;
+
+  User.findByIdAndUpdate(userId, { avatar }, { new: true, runValidators: true })
+    .then((user) => res.send({ newAvatar: user }))
+    .catch(() => res.status(500).send({ message: 'Произошла ошибка' }));
+};
+
+module.exports.login = (req, res) => {
+  const { email, password } = req.body;
+
+  return User.findUserByCredentials(email, password)
+    .then((user) => {
+      const token = jwt.sign({ _id: user._id }, JWT_SECRET_KEY, { expiresIn: '7d' });
+      res.cookie('jwt', token, {
+        maxAge: 3600000 * 24 * 7,
+        httpOnly: true,
+      });
+      res.send({ message: 'Авторизация прошла успешно' });
+    })
+    .catch((err) => {
+      res.status(err.statusCode || 500).send({ message: 'Что-то пошло не так', err: err.message });
```

Comment:
Можно лучше: не отправлять ошибку при ответе с кодом 500.
Когда происходит ошибка на сервере, то в теле ошибки обычно содержится stack trace ошибки + какая-то детальная информация. Получив эту информацию, пользователь никак не сможет её исправить.
В боевых проектах такие ошибки отлавливаются сервесами ошибок типа Sentry, Honeybadger.

```suggestion
      res.status(err.statusCode || 500).send({ message: 'Что-то пошло не так' });
```

Плюс можно унифицировать ответ для 500 ошибки. Чтобы в любом контроллере при ответе с 500 отправлялось одно и то же тело ответа.

Reply:
Можно лучше: унифицировать ответ для 500 ошибки.
Это актуально именно для 500, потому что ситуаций, когда возникает ошибка на сервере очень много. Для всех остальных лучше отправлять тот текст, который передается. Те:
```suggestion
      const statusCode = err.statusCode || 500;
      const message = statusCode === 500 ? 'Произошла ошибка' : err.message;
      res.status(statusCode).send({ message })
```